patches/comfyui_3d_pack.diff

```diff
@@ -1,3 +1,218 @@
++++ b/Gen_3D_Modules/Unique3D/custum_3d_diffusion/custum_pipeline/unifield_pipeline_img2img.py
+@@ -13,10 +13,11 @@
+ # limitations under the License.
+ # modified by Wuvin
+ 
+ from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+ 
++import comfy.utils
+ import numpy as np
++import PIL
+ import torch
+ 
+ from diffusers import AutoencoderKL, UNet2DConditionModel, StableDiffusionImageVariationPipeline
+@@ -96,10 +97,97 @@ class StableDiffusionImageCustomPipeline(
+ 
+         return image_embeddings
+ 
++    # adapted from https://github.com/huggingface/diffusers/blob/v0.12.0/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L471
++    def get_timesteps(self, num_inference_steps, new_steps, denoise):
++        # get the original timestep using init_timestep
++        init_timestep = min(int(num_inference_steps * denoise),
++                            num_inference_steps)
++        if init_timestep == 0:
++            raise ValueError(f"denoise {denoise} is too small.")
++        t_start = max(new_steps - num_inference_steps, 0)
++        timesteps = self.scheduler.timesteps[t_start:]
++
++        return timesteps, new_steps - t_start
++
++    # adapted from https://github.com/huggingface/diffusers/blob/v0.12.0/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L480
++    def prepare_latents(self,
++                        image,
++                        height,
++                        width,
++                        timestep,
++                        batch_size,
++                        num_images_per_prompt,
++                        dtype,
++                        device,
++                        generator=None):
++        batch_size = batch_size * num_images_per_prompt
++        if image is None:
++            return super().prepare_latents(
++                    batch_size,
++                    self.unet.config.out_channels,
++                    height,
++                    width,
++                    dtype,
++                    device,
++                    generator,
++                )
++
++        if not isinstance(image, (torch.Tensor, PIL.Image.Image, list)):
++            raise ValueError(
++                f"`image` has to be of type `torch.Tensor`, `PIL.Image.Image` or list but is {type(image)}"
++            )
++
++        image = image.permute(0, 3, 1, 2)
++        image = image.to(device=device, dtype=dtype)
++        if isinstance(generator, list) and len(generator) != batch_size:
++            raise ValueError(
++                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
++                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
++            )
++
++        if isinstance(generator, list):
++            init_latents = [
++                self.vae.encode(image[i:i + 1]).latent_dist.sample(
```

Comment:
separate encoding from this function and pass the latents, making the prepare_latents function minimally different from the original one